src/parser.rs

```diff
@@ -699,59 +680,53 @@ impl Parser {
-                Token::Word(k) if k.keyword == "IN" => Ok(Self::BETWEEN_PREC),
-                Token::Word(k) if k.keyword == "BETWEEN" => Ok(Self::BETWEEN_PREC),
-                Token::Word(k) if k.keyword == "LIKE" => Ok(Self::BETWEEN_PREC),
-                Token::Eq | Token::Lt | Token::LtEq | Token::Neq | Token::Gt | Token::GtEq => {
-                    Ok(20)
-                }
-                Token::Pipe => Ok(21),
-                Token::Caret => Ok(22),
-                Token::Ampersand => Ok(23),
-                Token::Plus | Token::Minus => Ok(Self::PLUS_MINUS_PREC),
-                Token::Mult | Token::Div | Token::Mod | Token::StringConcat => Ok(40),
-                Token::DoubleColon => Ok(50),
+        let token = self.peek_token();
+        debug!("get_next_precedence() {:?}", token);
+        match token {
+            Token::Word(w) if w.keyword == "OR" => Ok(5),
+            Token::Word(w) if w.keyword == "AND" => Ok(10),
+            Token::Word(w) if w.keyword == "NOT" => match self.peek_nth_token(1) {
+                // The precedence of NOT varies depending on keyword that
+                // follows it. If it is followed by IN, BETWEEN, or LIKE,
+                // it takes on the precedence of those tokens. Otherwise it
+                // is not an infix operator, and therefore has zero
+                // precedence.
+                Token::Word(w) if w.keyword == "IN" => Ok(Self::BETWEEN_PREC),
+                Token::Word(w) if w.keyword == "BETWEEN" => Ok(Self::BETWEEN_PREC),
+                Token::Word(w) if w.keyword == "LIKE" => Ok(Self::BETWEEN_PREC),
                 _ => Ok(0),
-            }
-        } else {
-            Ok(0)
+            },
+            Token::Word(w) if w.keyword == "IS" => Ok(17),
+            Token::Word(w) if w.keyword == "IN" => Ok(Self::BETWEEN_PREC),
+            Token::Word(w) if w.keyword == "BETWEEN" => Ok(Self::BETWEEN_PREC),
+            Token::Word(w) if w.keyword == "LIKE" => Ok(Self::BETWEEN_PREC),
+            Token::Eq | Token::Lt | Token::LtEq | Token::Neq | Token::Gt | Token::GtEq => Ok(20),
+            Token::Pipe => Ok(21),
+            Token::Caret => Ok(22),
+            Token::Ampersand => Ok(23),
+            Token::Plus | Token::Minus => Ok(Self::PLUS_MINUS_PREC),
+            Token::Mult | Token::Div | Token::Mod | Token::StringConcat => Ok(40),
+            Token::DoubleColon => Ok(50),
+            _ => Ok(0),
         }
     }
 
     /// Return the first non-whitespace token that has not yet been processed
     /// (or None if reached end-of-file)
-    pub fn peek_token(&self) -> Option<Token> {
+    pub fn peek_token(&self) -> Token {
         self.peek_nth_token(0)
     }
 
     /// Return nth non-whitespace token that has not yet been processed
-    pub fn peek_nth_token(&self, mut n: usize) -> Option<Token> {
+    pub fn peek_nth_token(&self, mut n: usize) -> Token {
         let mut index = self.index;
         loop {
             index += 1;
```

Comment:
This index += 1 I think would be better at the and of the `loop` so it can be just `self.tokens.get(index)`, or convert to a normal for loop?
(just looking at the code surrounding the changes)

Reply:
There's a `continue` later on, so moving it below won't work. If you have suggestions for improvements, feel free to raise separate PRs.